### PR TITLE
Vqgan

### PR DESCRIPTION
**Pull request by arthurdrake1**
Adds the VQGAN training and evaluation with some tuned hyperparameters and using an [Online Clustered Codebook](https://github.com/lyndonzheng/CVQ-VAE) to promote 100% codebook utilization. 

Also includes three new functions in `transforms.py` that could be useful for most/all EngiOpt models: 

- `resize_to` provides a standardized way of resizing tensors, i.e., to 128x128 prior to model training and vice versa
- `normalize` to normalize the conditions for zero mean and unit std
- `drop_constant` to remove any condition columns that remain constant throughout the data (i.e., `overhang_constrant` = 0 for all of beams2d)

The latter two are optional and can be disabled in the args for VQGAN.

### CODE REVIEW

File: engiopt/vqgan/evaluate_vqgan.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    model.eval()  # Set to evaluation mode
+    model.to(device)
+
+    ### Set up testing conditions ###
+    _, sampled_conditions, sampled_designs_np, _ = sample_conditions(
+        problem=problem, n_samples=args.n_samples, device=device, seed=seed
+    )
+
+    # Clean up conditions based on model training settings and convert back to tensor
+    sampled_conditions_new = sampled_conditions.select(range(len(sampled_conditions)))
```
Comment (ffelten) [inline engiopt/vqgan/evaluate_vqgan.py]: I don't get what this does

File: pyproject.toml [hunk 2]
```diff
@@ -132,7 +134,6 @@ ignore = [
   "S607",     # start-process-with-partial-path
   "T201",     # print
   "TRY003",     # print
-  "PD901",     # avoid using df to name dataframe
```
Comment (ffelten) [inline pyproject.toml]: Why removing this?

File: engiopt/vqgan/vqgan.py [hunk 3]
```diff
@@ -0,0 +1,1900 @@
...
+        encoder_start_resolution: int,
+        encoder_attn_resolutions: tuple[int, ...],
+        encoder_num_res_blocks: int,
+        image_channels: int,
+        latent_dim: int,
+    ):
+        super().__init__()
+        channels = encoder_channels
+        resolution = encoder_start_resolution
+        layers = [nn.Conv2d(image_channels, channels[0], 3, 1, 1)]
```
Comment (ffelten) [inline engiopt/vqgan/vqgan.py]: I would use named args for the Conv2d, it is not straightforward to understand this line without them

File: engiopt/vqgan/vqgan.py [hunk 4]
```diff
@@ -0,0 +1,1900 @@
...
+        decoder_attn_resolutions: tuple[int, ...],
+        decoder_num_res_blocks: int,
+        image_channels: int,
+        latent_dim: int,
+    ):
+        super().__init__()
+        in_channels = decoder_channels[0]
+        resolution = decoder_start_resolution
+        layers = [
+            nn.Conv2d(latent_dim, in_channels, 3, 1, 1),
```
Comment (ffelten) [inline engiopt/vqgan/vqgan.py]: same here

File: engiopt/vqgan/vqgan.py [hunk 5]
```diff
@@ -0,0 +1,1900 @@
...
+                if resolution in decoder_attn_resolutions:
+                    layers.append(NonLocalBlock(in_channels))
+
+            if i != 0:
+                layers.append(UpSampleBlock(in_channels))
+                resolution *= 2
+
+        layers.append(GroupNorm(in_channels))
+        layers.append(Swish())
+        layers.append(nn.Conv2d(in_channels, image_channels, 3, 1, 1))
```
Comment (ffelten) [inline engiopt/vqgan/vqgan.py]: and here :)

File: engiopt/vqgan/vqgan.py [hunk 6]
```diff
@@ -0,0 +1,1900 @@
...
+        """Forward pass with optional CVQGAN adapter."""
+        return self.model(x)
+
+
+class ScalingLayer(nn.Module):
+    """Channel-wise affine normalization used by LPIPS."""
+
+    def __init__(self):
+        super().__init__()
+        self.register_buffer("shift", th.tensor([-0.030, -0.088, -0.188])[None, :, None, None])
```
Comment (ffelten) [inline engiopt/vqgan/vqgan.py]: These constants are coming from LPIPS?

File: engiopt/vqgan/vqgan.py [hunk 7]
```diff
@@ -0,0 +1,1900 @@
...
+        n_head (int): Number of attention heads in the Transformer.
+        n_embd (int): Embedding dimension in the Transformer.
+        dropout (float): Dropout rate in the Transformer.
+        bias (bool): If True, use bias terms in the Transformer layers.
+    """
+
+    def __init__(  # noqa: PLR0913
+        self,
+        *,
+        conditional: bool = True,
```
Comment (ffelten) [inline engiopt/vqgan/vqgan.py]: Is this correct python? I thought you can only put arguments with default values after arguments without default

File: engiopt/vqgan/vqgan.py [hunk 8]
```diff
@@ -0,0 +1,1900 @@
...
+        encoded = self.model(x)
+        s = encoded.shape
+        return encoded.view(s[0], s[1] // self.c_feature_map_dim**2, self.c_feature_map_dim, self.c_feature_map_dim)
+
+
+class Decoder(nn.Module):
+    """Decoder module for VQGAN Stage 1.
+
+    Consists of a series of convolutional, residual, and attention blocks arranged using the provided arguments.
+    The number of upsample blocks is determined by the length of the decoder channels tuple minus one.
```
[the remark this answers is not in the document]
Comment (arthurdrake1) [reply engiopt/vqgan/vqgan.py]: Same as above, architectural choice in original repo

File: engiopt/vqgan/vqgan.py [hunk 9]
```diff
@@ -0,0 +1,1900 @@
...
+    n_conds = len(conditions)
+    args.cond_dim = n_conds
+    condition_tensors = [training_ds[key][:] for key in conditions]
+
+    # Move to device only here
+    th_training_ds = th.utils.data.TensorDataset(
+        th.as_tensor(training_ds["optimal_upsampled"][:]).to(device),
+        *[th.as_tensor(training_ds[key][:]).to(device) for key in conditions],
+    )
+    dataloader_0 = th.utils.data.DataLoader(
```
[the remark this answers is not in the document]
Comment (arthurdrake1) [reply engiopt/vqgan/vqgan.py]: The different model stages could have different batch sizes

File: engiopt/vqgan/vqgan.py [hunk 10]
```diff
@@ -0,0 +1,1900 @@
...
+    if args.track:
+        wandb.init(
+            project=args.wandb_project,
+            entity=args.wandb_entity,
+            config=vars(args),
+            save_code=True,
+            name=run_name,
+            dir="./logs/wandb",
+        )
+        wandb.define_metric("0_step", summary="max")
```
[the remark this answers is not in the document]
Comment (arthurdrake1) [reply engiopt/vqgan/vqgan.py]: Yes in this special multi-stage training case, to have separate training step counts for each one.

**Comment by arthurdrake1** [comment]
@ffelten Going to make the minor edits now. I will also move all the "helper modules" like nanogpt into their own file. As for separating out the different training phases, this would be more complicated as the current logic is built on the single script implementation, i.e., easily reusing the trained vqgan and cvqgan to then train the transformer.

File: engiopt/vqgan/evaluate_vqgan.py (hunk 1 above)
Comment (ffelten) [reply engiopt/vqgan/evaluate_vqgan.py]: Fair! The name is a bit cryptic then, though.

File: engiopt/vqgan/vqgan.py [hunk 11]
```diff
@@ -0,0 +1,1213 @@
...
+    cond_feature_map_dim: int = 4
+    """feature map dimension for the CVQGAN encoder output"""
+    batch_size_cvqgan: int = 16
+    """size of the batches for CVQGAN"""
+    n_epochs_cvqgan: int = 1000  # Default: 1000
+    """number of epochs of CVQGAN training"""
+    cond_lr: float = 2e-4  # Default: 2e-4
+    """learning rate for CVQGAN"""
+    latent_size: int = 16
+    """size of the latent feature map (automatically determined later)"""
```
Comment (ffelten) [inline engiopt/vqgan/vqgan.py]: Why is this determined later? Also, is this "automated determination" reflected in the wandb config (hyperparams) ?

**Review by ffelten — APPROVED** [review]
Some minor stuff. I did not read all the details but it looks clean enough to me!

File: engiopt/vqgan/vqgan.py (hunk 11 above)
Comment (arthurdrake1) [reply engiopt/vqgan/vqgan.py]: The latent feature map dimension is dictated by the number of encoder and decoder layers, so in practice the user has to increase or decrease those to change `latent_size`. Number of image channels is more straightforward, we just update that based on the number of channels in the data (always 1 in our case, for now). 

I checked the wandb config for a test run I did just now with bogus values (`--latent_size 123123` and `--image_channels 123123`). They remained at 16 and 1 respectively in the hyperparams list on wandb.

Comment (arthurdrake1) [reply engiopt/vqgan/vqgan.py]: @ffelten Upon further review of the code, I don't think these args are needed. They're just calculated as normal variables and override any actual user inputs as I determined above. Should be good to go now.
